ospfd/ospf_gr.c

```diff
@@ -0,0 +1,480 @@
+
+	options = LSA_OPTIONS_GET(oi->area);
+	options |= LSA_OPTIONS_NSSA_GET(oi->area);
+	options |= OSPF_OPTION_O;
+
+	lsa_type = OSPF_OPAQUE_LINK_LSA;
+	tmp = SET_OPAQUE_LSID(OPAQUE_TYPE_GRACE_LSA, 0);
+	lsa_id.s_addr = htonl(tmp);
+
+	/* Set opaque-LSA header fields. */
+	lsa_header_set(s, options, lsa_type, lsa_id, oi->ospf->router_id);
+
+	/* Set opaque-LSA body fields. */
+	ospf_gr_lsa_body_set(&oi->ospf->gr_info, s, oi);
+
+	/* Set length. */
+	length = stream_get_endp(s);
+	lsah->length = htons(length);
+
+	/* Now, create an OSPF LSA instance. */
+	if ((new = ospf_lsa_new_and_data(length)) == NULL) {
+		zlog_warn("ospf_gr_lsa_new: ospf_lsa_new_and_data() failed");
+		stream_free(s);
+		goto out;
+	}
+
+	zlog_debug("LSA[Type%d:%s]: Create an Opaque-LSA/GR instance", lsa_type,
+		   inet_ntoa(lsa_id));
+
+	new->area = oi->area;
+	new->oi = oi;
+	SET_FLAG(new->flags, OSPF_LSA_SELF);
+	memcpy(new->data, lsah, length);
+	stream_free(s);
+
+out:
+	return new;
+}
+
+static struct ospf_lsa *ospf_gr_lsa_originate(struct ospf_interface *oi,
+					      struct ospf_lsa *old)
+{
+	struct ospf_lsa *new = NULL;
+	struct ospf_area *area;
+
+	if (oi->state == ISM_Down) {
+		zlog_debug(
+			"%s:Originating Grace LSA on Down interface %s (abort)",
+			__func__, IF_NAME(oi));
+		goto out;
+	}
+
+	/* Create new Grace-LSA instance. */
+	if ((new = ospf_gr_lsa_new(oi)) == NULL) {
+		zlog_warn("ospf_gr_lsa_generate: ospf_gr_lsa_new() failed");
+		goto out;
+	}
+
+	/* Adjust the sequence number */
+	if (!old) {
+		/* Find the old LSA and increase the seq. */
+		area = ospf_area_get(oi->ospf, oi->ospf->router_id);
+		old = ospf_lsa_lookup_by_header(area, new->data);
+		if (old)
+			new->data->ls_seqnum = lsa_seqnum_increment(old);
+	}
+
+	/* Install this LSA into LSDB. */
+	if (ospf_lsa_install(oi->ospf, oi, new) == NULL) {
+		zlog_warn("ospf_gr_lsa_generate: ospf_lsa_install() failed");
+		ospf_lsa_unlock(&new);
+		goto out;
+	}
+
+	/* Update new LSA origination count. */
+	if (!old)
+		oi->ospf->lsa_originate_count++;
+
+	/* Flood the LSA through out the interface */
+	ospf_flood_through_interface(oi, NULL, new);
+
+out:
+	return new;
+}
+
+static struct ospf_lsa *ospf_gr_lsa_refresh(struct ospf_lsa *lsa)
+{
+	struct ospf *ospf = lsa->oi->ospf;
+	bool force_org = false;
+
+	if (!IS_GR_LSA(lsa->data) || !CHECK_FLAG(lsa->flags, OSPF_LSA_SELF))
+		return NULL;
+
+	if (!ospf->gr_info.restart_support) {
+		/* Seems to be a leftover, flush it away. */
+		lsa->data->ls_age = htons(OSPF_LSA_MAXAGE);
+	} else if (ospf->gr_info.prepare_running) {
+		/* Prepare running, refresh this LSA */
+		force_org = true;
+	} else {
+		lsa->data->ls_age = htons(OSPF_LSA_MAXAGE);
+	}
+
+	/* If the lsa's age reached to MaxAge, start flushing procedure. */
+	if (IS_LSA_MAXAGE(lsa) && !force_org) {
+		ospf_opaque_lsa_flush_schedule(lsa);
+	} else {
+		ospf_gr_init_period(&lsa->oi->ospf->gr_info);
+		ospf_gr_lsa_originate(lsa->oi, lsa);
+	}
+
+	return NULL;
```

Comment:
That is necessary because this is a callback (with a specified signature) used in e.g. the `functab` related code. I just noticed that again after changing the code :).